toolchain/semantics/semantics_ir_factory.cpp

```diff
@@ -165,91 +65,210 @@ static auto GetBinaryOp(TokenKind kind) -> Semantics::BinaryOperator::Op {
-    Semantics::NodeId target_id) {
-  CARBON_CHECK(parse_tree().node_kind(node) == ParseNodeKind::InfixOperator());
-
-  auto token = parse_tree().node_token(node);
-  auto token_kind = tokens_->GetKind(token);
-  auto op = GetBinaryOp(token_kind);
-
-  auto rhs_id = next_id();
-  auto lhs_id = next_id();
-  nodes.push_back(semantics_.nodes_.Store(
-      Semantics::BinaryOperator(node, target_id, op, lhs_id, rhs_id)));
-  auto subtree = ParseSubtreeConsumer::ForParent(parse_tree(), node);
-  TransformExpression(nodes, subtree.RequireConsume(), rhs_id);
-  TransformExpression(nodes, subtree.RequireConsume(), lhs_id);
-}
-
-/*
-auto SemanticsIRFactory::TransformParameterList(ParseTree::Node node)
-    -> llvm::SmallVector<Semantics::PatternBinding, 0> {
-  CARBON_CHECK(parse_tree().node_kind(node) ==
-ParseNodeKind::ParameterList());
-
-  auto subtree = ParseSubtreeConsumer::ForParent(parse_tree(), node);
-  RequireNodeEmpty(subtree.RequireConsume(ParseNodeKind::ParameterListEnd()));
-
-  llvm::SmallVector<Semantics::PatternBinding, 0> params;
-  if (auto first_param_node =
-          subtree.TryConsume(ParseNodeKind::PatternBinding())) {
-    params.push_back(TransformPatternBinding(*first_param_node));
-
-    while (auto comma_node =
-               subtree.TryConsume(ParseNodeKind::ParameterListComma())) {
-      RequireNodeEmpty(*comma_node);
-      params.push_back(TransformPatternBinding(
-          subtree.RequireConsume(ParseNodeKind::PatternBinding())));
+// bazel run //toolchain/driver:carbon dump semantics-ir
+// testdata/function/basic.carbon
+void SemanticsIRFactory::Build() {
+  // Silence "unused" build warning.
+  llvm::SmallVector<TraversalStackEntry, 0> node_stack;
+  auto range = parse_tree().postorder();
+  for (auto it = range.begin();; ++it) {
+    auto parse_node = *it;
+    int32_t subtree_size = parse_tree().node_subtree_size(parse_node);
+    switch (auto parse_kind = parse_tree().node_kind(parse_node)) {
+      case ParseNodeKind::CodeBlock(): {
+        // Just merges children.
+        llvm::SmallVector<Semantics::NodeRef, 0> sem_ir;
+        while (subtree_size > 1) {
+          subtree_size -=
+              parse_tree().node_subtree_size(node_stack.back().parse_node);
+          sem_ir.insert(sem_ir.begin(), node_stack.back().sem_ir.begin(),
+                        node_stack.back().sem_ir.end());
+          node_stack.pop_back();
+        }
+        node_stack.push_back(
+            TraversalStackEntry(parse_node, std::move(sem_ir)));
+        break;
+      }
+      case ParseNodeKind::DeclaredName(): {
+        RequireNodeEmpty(parse_node, parse_kind, subtree_size);
+        // DeclaredNames will be acted upon as part of the declaring construct.
+        node_stack.push_back(TraversalStackEntry(parse_node));
+        break;
+      }
+      case ParseNodeKind::FunctionDeclaration(): {
+        // Currently we only have definitions, so this is a CodeBlock.
+        subtree_size -=
```

Comment:
I wonder whether it's worth adding a wrapper around the `int32` here with an `operator-=` that `CARBON_CHECK`s the resulting integer is `>= 1`. Or maybe even a combined "pop last node from the node stack and reduce the subtree size" operation would make sense.

Reply:
Done